### ISSUE TITLE
cross check specification: handling old fmu's uploaded before cross check rules.

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/fmi-standard#336]
**Reported by karl&#46;wernersson on 12 May 2015 17:15 UTC**
Hi

As one of the earliest up-loader of fmu's to the svn repository, I have several FMU's uploaded before any cross check rules were specified. These of course does not follow the latest rules. (Specifically FMI1 Dymola 2013 FD01).

I am also not alone in having a (sub)set of uploaded FMU's that are not compliant with the latest rules (or contains other errors).

These old fmu's is a problem for Importers since there might be cases where it is impossible to provide import result due to missing files in the export result. And it is not defined for that case whether to  ignore, reject or fail the FMU.

However the rules also forbid an exporter to change the submission since this might change already finalized import results.

In the future new cross check rules may also make all current uploads non compliant with the future rules. (I hope not). This will then once again create allot of messy work for all fmu submitters.

I have two proposals to solve the problem.

1. Allow fmu submitter to clean up and resubmit non compliant fmu's. (If one has a good script it is easy to generate everything new according to the latest rules.

Problems that could occur: new FMU's might not be consistent whit the old due to missing source model etc., that tool version could have produced FMU's that are no longer possible to commit e.g. a more advanced compliance checker or commit script finding errors with the FMU that was not detected at the original time of submission.

2. Marking fmu's exports folders that are not compliant with an "obsolete" or "notCompliantWithLatestRules" file. 
This would allow importers to easily ignore FMU's that a tool vendor is not interested in keeping updated. It could also possible be shown in the cross check matrix encouraging tool vendors to either update their FMU submission or submitting new FMU's from a later tool version.

/Karl Wernersson

----

Migrated-From: https://trac.fmi-standard.org/ticket/336

**Comment by modelica-trac-importer** [comment]
**Comment by andreas&#x2e;junghanns on 13 May 2015 13:17 UTC**
I agree with both suggestions:

1) We missed this part in the rules documents and never explicitly state, that we allow fixing of "old" FMUs. But that was discussed and I do not remember opposition. I would even go as far as saying: non-complying, pre-rule-submissions NEED fixing. That is in the spirit of the overall idea. We just have to prevent people from changing their "normal" and new submissions again and again.  I will ask Torsten to put this on the agenda for next Steering Committee meeting and there I will propose that we allow changes as you propose.

2) Totally agree. This is a backward compatible change, so I suggest we add this right away. As "obsolete" is more vague than "notCompliantWithLatestRules", I suggest the later to be used. I'll add that as well on the agenda as a second point.

**Comment by modelica-trac-importer** [comment]
**Modified by andreas&#x2e;junghanns on 13 May 2015 13:23 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by cbertsch on 10 Jun 2015 19:23 UTC**
Regarding 1.: What about FMUs that do not violate the cross check rules but the FMI standard itself?
Are they allowed to be removed? As those FMUs have been generated with released tool versions, I would l like to keep them, because the "red" entries in the cross-check table might contain valuable information for the end user. So I would prefer only 2., where old commits could be marked as obsolete, but would not be removed.

**Comment by modelica-trac-importer** [comment]
**Comment by andreas&#x2e;junghanns on 10 Jun 2015 19:54 UTC**
Replying to [comment:3 cbertsch]:
> Regarding 1.: What about FMUs that do not violate the cross check rules but the FMI standard itself?
> Are they allowed to be removed? As those FMUs have been generated with released tool versions, I would l like to keep them, because the "red" entries in the cross-check table might contain valuable information for the end user. So I would prefer only 2., where old commits could be marked as obsolete, but would not be removed.

Indeed, I think I was not clear and re-reading my earlier sentences, this can be misunderstood. So to be clear: Changing submissions means changing/adding/renaming any file, except the FMU. The FMU is a fix point: it depends on the tool and version specified - how could that change to warrant changes?

@Karl: Do we agree on that?

**Comment by modelica-trac-importer** [comment]
**Comment by karl&#46;wernersson on 16 Jun 2015 10:09 UTC**
Yes, I can agree on that as long as the following behavior seems reasonable.

I would leave old FMU's as is without changes and marked as "notCompliantWithLatestRules", as there are never versions of my tool with FMU's uploaded that follow the rules.

If one have a **good reason** and request an update to my old FMU's on an old tool version, I would, depending on what needs fixing and if possible, upload a new batch parallel to the previous on that tool release instead of locally fixing files

**Comment by modelica-trac-importer** [comment]
**Comment by andreas&#46;junghanns on 19 Jun 2015 13:12 UTC**
Please check my changes to the documents:
```
...trunk/FMI_Cross_Check/FMI_Cross_Check_Rules_v3.1_2015_06.odt
```
and
```
...trunk/FMI_Cross_Check/FMI_FMI_Cross_Check_Implementation_Notes_v3.1_2015_06.odt
```

(MS Word supports that format in the later versions, so I am told.)

**Comment by modelica-trac-importer** [comment]
**Comment by anonymous on 3 Jul 2015 11:17 UTC**
Hi

I added some more info in the cross check rules regarding this issue, mainly about adding a new cell in the table for this result. I chose the color blue.

The updated file is called
FMI_Cross_Check_Rules_v3.1_2015_07.odt
I also added a word and pdf version with change remark for those interested in seeing the changes and if we need to revert some of them

/Kalle

**Comment by modelica-trac-importer** [comment]
**Comment by andreas&#x2e;junghanns on 3 Jul 2015 12:37 UTC**
Replying to [comment:7 anonymous]:
> Hi
> 
> I added some more info in the cross check rules regarding this issue, mainly about adding a new cell in the table for this result. I chose the color blue.

Since adding the post-commit-hook we should add to this problem.
Since we allow old FMU submits to be fixed according to the XC rules in the next version,
the problems we have should be reduced, or maybe removed entirely.

Do we really want to treat these, hopefully soon-to-be-extinct, cases explicitly in the rules?

Pro: we might tighten the rules further in the future leading to more non-compliant cases that we might want to single out.

Con: if we ever introduce such drastic changes to the rules, we could introduce how to deal with those cases then (and much more accurately then today when we have to anticipate the resulting problems).

Do we know how many cases we currently have that fail the XC rules (== the post-commit-hook)?

> The updated file is called
> FMI_Cross_Check_Rules_v3.1_2015_07.odt
> I also added a word and pdf version with change remark for those interested in seeing the changes and if we need to revert some of them

I edited the 2015_6.odt version to include your removal of the FMI 2.0 results - thank you for catching this.

In a comment, you raise an intersting question w.r.t to Refence FMUs:
```
Is this relevant? Do we have any reference FMUS? Is someone curently creating these references? How whould they be reported in the matrxi?
```
To answer your questions: Nobody produced those, even though several members initially proposed to do that. I still think this might be a good idea, but as far as I know, this is largely forgotten.

I took your comment and moved it also into the 2015_6.odt version.
Except for the last question (which I did not copy): Reference FMUs are treated by the script like any other export tool.


----


To summarise:
1. I agree with and added 1 of 3 of your proposals to what I consider the master document FMI_Cross_Check_Rules_v3.1_2015_06.odt (removal of obsolete comment).

2. We should pose the problem of still not existing reference FMUs to the group and if they agree, remove that section from the rules and declare the idea of Reference FMUs to be dead.

3. Depending on the discussion about the treatment of non-compliant results, we might have to add your proposal later (but I hope we will agree that this is a small, temporary problem that can be ignored until gone).

**Comment by modelica-trac-importer** [comment]
**Comment by izacharias on 8 Jul 2015 06:22 UTC**
In FMI_Cross_Check_Rules_v3.1_2015_06.odt in chapter 5, rule # 11, was adapted:

```
If a submission does not comply with the latest submission rules, a file named
“notCompliantWithLatestRules “.
```

# 11 is a rule for FMU importers. I think that the notCompliantWithLatestRules file must be added to the FMU’s Export folder, not to the test result folder, i.e. # 9 should be adapted.

**Comment by modelica-trac-importer** [comment]
**Comment by andreas&#46;junghanns on 8 Jul 2015 08:00 UTC**
Replying to [comment:9 izacharias]:
> In FMI_Cross_Check_Rules_v3.1_2015_06.odt in chapter 5, rule # 11, was adapted:
> 
> ```
> If a submission does not comply with the latest submission rules, a file named
> “notCompliantWithLatestRules “.
> ```
> 
> # 11 is a rule for FMU importers. I think that the notCompliantWithLatestRules file must be added to the FMU’s Export folder, not to the test result folder, i.e. # 9 should be adapted.

Right, sorry. I fixed this and made sure the rule states that changes to submit files are allowed to comply to the new rules again and then to remove that file as well.
Thanks for catching this!!

**Comment by modelica-trac-importer** [comment]
**Comment by andreas&#46;junghanns on 8 Jul 2015 15:26 UTC**
I close this ticket, hoping to have addressed your problem. If not, feel free to reopen, thank you!
